File: web/src/__tests__/data/NoteIdSentinel.test.ts
import { describe, it, expect } from 'vitest'
import {
  newSentinelNoteId,
  isSentinelNoteId,
  originOfSentinel,
  NOTE_ID_SENTINEL_PREFIX,
} from '../../data/NoteIdSentinel'

describe('NoteIdSentinel', () => {
  it('newSentinelNoteId produces sentinel with correct prefix and origin code', () => {
    const id = newSentinelNoteId('paste')
    expect(id.startsWith(NOTE_ID_SENTINEL_PREFIX)).toBe(true)
    expect(id).toMatch(/^@paste_[a-f0-9]+$/i)
  })

  it('isSentinelNoteId recognizes sentinels and rejects real ids', () => {
    expect(isSentinelNoteId('@paste_abc123')).toBe(true)
    expect(isSentinelNoteId(newSentinelNoteId('split'))).toBe(true)
    expect(isSentinelNoteId('aE5eVp2p2uiusUXqVa8T')).toBe(false)
    expect(isSentinelNoteId(null)).toBe(false)
    expect(isSentinelNoteId(undefined)).toBe(false)
    expect(isSentinelNoteId('')).toBe(false)
  })

  it('originOfSentinel extracts origin code', () => {
    expect(originOfSentinel(newSentinelNoteId('paste'))).toBe('paste')
    expect(originOfSentinel(newSentinelNoteId('split'))).toBe('split')
    expect(originOfSentinel('@agent_xyz')).toBe('agent')
    expect(originOfSentinel('aE5eVp2p2uiusUXqVa8T')).toBeNull()
    expect(originOfSentinel(null)).toBeNull()
  })

  it('newSentinelNoteId generates distinct ids across calls', () => {
    const ids = Array.from({ length: 20 }, () => newSentinelNoteId('paste'))
    expect(new Set(ids).size).toBe(ids.length)
  })

  const origins: Array<'typed' | 'paste' | 'split' | 'agent' | 'directive' | 'surgical'> = [
    'typed', 'paste', 'split', 'agent', 'directive', 'surgical',
  ]
  it.each(origins)('origin %s round-trips through originOfSentinel', (origin) => {
    const id = newSentinelNoteId(origin)
    expect(originOfSentinel(id)).toBe(origin)
  })
})

File: app/src/test/java/org/alkaline/taskbrain/data/NoteIdSentinelTest.kt
package org.alkaline.taskbrain.data

import org.junit.Assert.*
import org.junit.Test

class NoteIdSentinelTest {
    @Test
    fun `new - produces sentinel with correct prefix and origin code`() {
        val id = NoteIdSentinel.new(NoteIdSentinel.Origin.PASTE)
        assertTrue("should start with @", id.startsWith("@"))
        assertTrue("should contain origin code", id.contains("paste_"))
        assertTrue("should have entropy after the code", id.length > "@paste_".length)
    }

    @Test
    fun `isSentinel - recognizes sentinels and rejects real ids`() {
        assertTrue(NoteIdSentinel.isSentinel("@paste_abc123"))
        assertTrue(NoteIdSentinel.isSentinel(NoteIdSentinel.new(NoteIdSentinel.Origin.SPLIT)))
        assertFalse(NoteIdSentinel.isSentinel("aE5eVp2p2uiusUXqVa8T")) // real Firestore id
        assertFalse(NoteIdSentinel.isSentinel(null))
        assertFalse(NoteIdSentinel.isSentinel(""))
    }

    @Test
    fun `originOf - extracts origin code`() {
        assertEquals("paste", NoteIdSentinel.originOf(NoteIdSentinel.new(NoteIdSentinel.Origin.PASTE)))
        assertEquals("split", NoteIdSentinel.originOf(NoteIdSentinel.new(NoteIdSentinel.Origin.SPLIT)))
        assertEquals("agent", NoteIdSentinel.originOf("@agent_xyz"))
        assertNull(NoteIdSentinel.originOf("aE5eVp2p2uiusUXqVa8T"))
        assertNull(NoteIdSentinel.originOf(null))
    }

    @Test
    fun `new - generates distinct ids on each call`() {
        val ids = (1..20).map { NoteIdSentinel.new(NoteIdSentinel.Origin.PASTE) }
        assertEquals("all ids should be distinct", ids.size, ids.toSet().size)
    }

    @Test
    fun `all origins round-trip through originOf`() {
        for (origin in NoteIdSentinel.Origin.values()) {
            val id = NoteIdSentinel.new(origin)
            assertEquals(origin.code, NoteIdSentinel.originOf(id))
        }
    }
}
